substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,545 @@
+                aMethod.registerAsImplementationInvoked(null);
+
+                if (!aMethod.isStatic() && args[0].isConstant()) {
+                    AnalysisType receiverType = (AnalysisType) StampTool.typeOrNull(args[0]);
+                    if (receiverType != null) {
+                        receiverType.registerAsInHeap();
+                    }
+                }
+            }
+            return InlineInfo.createStandardInlineInfo(callee);
+        } else {
+            return null;
+        }
+    }
+
+    static boolean inliningBeforeAnalysisNotAllowed(GraphBuilderContext b, ResolvedJavaMethod callee, ResolvedJavaMethod caller) {
+        return b.parsingIntrinsic() ||
+                        GuardedAnnotationAccess.isAnnotationPresent(callee, NeverInline.class) || GuardedAnnotationAccess.isAnnotationPresent(callee, NeverInlineTrivial.class) ||
+                        GuardedAnnotationAccess.isAnnotationPresent(callee, Uninterruptible.class) || GuardedAnnotationAccess.isAnnotationPresent(caller, Uninterruptible.class) ||
+                        GuardedAnnotationAccess.isAnnotationPresent(callee, RestrictHeapAccess.class) || GuardedAnnotationAccess.isAnnotationPresent(caller, RestrictHeapAccess.class) ||
+                        /*
+                         * Canonicalization during inlining folds to a constant in analysis, but not
+                         * for
+                         * com.oracle.svm.hosted.image.NativeImageCodeCache.buildRuntimeMetadata.
+                         * Either we need to re-use the analysis graphs or we have to apply the same
+                         * canonicalizations for buildRuntimeMetadata.
+                         */
+                        GuardedAnnotationAccess.isAnnotationPresent(caller, DeoptTest.class) ||
+                        /*
+                         * In Java 11 java.lang.invoke.InvokeDynamic for string concatenation
+                         * produces bci duplication that can lead to inconsistent inlining results.
+                         */
+                        filterStringConcatGeneratedMethods(callee) ||
+                        b.getDepth() > NativeImageInlineDuringParsingPlugin.Options.InlineBeforeAnalysisMaxDepth.getValue(b.getOptions()) ||
+                        isRecursiveCall(b, callee) ||
+                        /*
+                         * We don't want to process invokes if bci is not unique.
+                         */
+                        !b.bciUnique();
+    }
+
+    private static boolean isRecursiveCall(GraphBuilderContext b, ResolvedJavaMethod callee) {
+        return b.getCallingContext().stream().map(Pair::getLeft).anyMatch(caller -> caller.equals(callee));
+    }
+
+    private static boolean filterStringConcatGeneratedMethods(ResolvedJavaMethod method) {
+        return method.format("%H.%n").equals("java.lang.String.valueOf") || method.format("%H.%n").equals("java.lang.StringConcatHelper.prepend") ||
```

Comment:
This kind of code is OK during testing, however, it is absolutely not OK in production due to style, performance, and correctness issues. This kind of code, makes people feel bad about the whole approach and the reviewer gets serious doubts about the credibility of the whole approach.

Reply:
Yes, we cannot really merge that. But the proper fix should be really simple: while we are in a graph builder plugin (or any kind of intrinsic context), `BytecodeParser.bciUnique` needs to return `false` because any plugin can lead to more than one "replaced invoke" being produced.

Reply:
I added fix for this and removed method filterStringConcatGeneratedMethods.